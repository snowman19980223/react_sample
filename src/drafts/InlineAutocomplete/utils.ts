import getCaretCoordinates from '@koddsson/textarea-caret'
import {Children, EventHandler, SyntheticEvent} from 'react'

import {Coordinates, ShowSuggestionsEvent, Suggestion, TextInputCompatibleChild, Trigger} from './types'

const singleWordTriggerTerminators = new Set([' ', '\n'])
const multiWordTriggerTerminators = new Set(['.', '\n'])

const isWhitespace = (char: string) => /\s/.test(char)

/**
 * Calculate whether or not suggestions should be shown based on the given state of the
 * input. If they should be shown, returns the show event.
 */
export const calculateSuggestionsQuery = (
  triggers: Array<Trigger>,
  text: string,
  caretLocation: number
): ShowSuggestionsEvent | null => {
  // Build backwards from the caret location until the most recent trigger character or terminator
  for (
    let i = caretLocation - 1, query = '', potentialTriggers = triggers;
    i >= 0 && potentialTriggers.length > 0;
    i--
  ) {
    const character = text[i]

    if (singleWordTriggerTerminators.has(character)) potentialTriggers = potentialTriggers.filter(t => t.multiWord)
    if (multiWordTriggerTerminators.has(character)) potentialTriggers = potentialTriggers.filter(t => !t.multiWord)

    for (const trigger of potentialTriggers.filter(t => character === t.triggerChar)) {
      // Trigger chars must always be preceded by whitespace or be the first character in the input,
      // and even a multi-word query cannot start with whitespace
      if ((i === 0 || isWhitespace(text[i - 1])) && !isWhitespace(query[0])) return {trigger, query}

      potentialTriggers = potentialTriggers.filter(t => t !== trigger)
    }

    query = character + query
  }

  return null
}

/**
 * Obtain the coordinates (px) of the bottom left of a character in an input, relative to the
 * top-left corner of the input itself.
 * @param input The target input element.
 * @param index The index of the character to calculate for.
 * @param adjustForScroll Control whether the returned value is adjusted based on scroll position.
 */
export const getCharacterCoordinates = (
  input: HTMLTextAreaElement | HTMLInputElement | null,
  index: number,
  adjustForScroll = true
): Coordinates => {
  if (!input) return {top: 0, left: 0}

  // word-wrap:break-word breaks the getCaretCoordinates calculations (a bug), and word-wrap has
  // no effect on input element anyway
  if (input instanceof HTMLInputElement) input.style.wordWrap = ''

  let coords = getCaretCoordinates(input, index)

  // The library calls parseInt on the computed line-height of the element, failing to account for
  // the possibility of it being 'normal' (another bug). In that case, fall back to a rough guess
  // of 1.2 based on MDN: "Desktop browsers use a default value of roughly 1.2".
  if (isNaN(coords.height)) coords.height = parseInt(getComputedStyle(input).fontSize) * 1.2

  // Sometimes top is negative, incorrectly, because of the wierd line-height calculations around
  // border-box sized single-line inputs.
  coords.top = Math.abs(coords.top)

  // For some single-line inputs, the rightmost character can be accidentally wrapped even with the
  // wordWrap fix above. If this happens, go back to the last usable index
  let adjustedIndex = index
  while (input instanceof HTMLInputElement && coords.top > coords.height) {
    coords = getCaretCoordinates(input, --adjustedIndex)
  }

  const scrollTopOffset = adjustForScroll ? -input.scrollTop : 0
  const scrollLeftOffset = adjustForScroll ? -input.scrollLeft : 0

  return {top: coords.top + coords.height + scrollTopOffset, left: coords.left + scrollLeftOffset}
}

/**
 * Obtain the coordinates of the bottom left of a character in an input relative to the top-left
 * of the page.
 * @param input The target input element.
 * @param index The index of the character to calculate for.
 */
export const getAbsoluteCharacterCoordinates = (
  input: HTMLTextAreaElement | HTMLInputElement | null,
  index: number
): Coordinates => {
  const {top: relativeTop, left: relativeLeft} = getCharacterCoordinates(input, index, true)
  const {top: viewportOffsetTop, left: viewportOffsetLeft} = input?.getBoundingClientRect() ?? {top: 0, left: 0}

  return {
    top: viewportOffsetTop + relativeTop,
    left: viewportOffsetLeft + relativeLeft
  }
}

export const getSuggestionValue = (suggestion: Suggestion): string =>
  typeof suggestion === 'string' ? suggestion : suggestion.value

export const getSuggestionKey = (suggestion: Suggestion): string =>
  typeof suggestion === 'string' ? suggestion : suggestion.key ?? suggestion.value

/**
 * Replace a section of a string.
 */
export const replaceSlice = (
  original: string,
  [startInclusive, endExclusive]: [number, number],
  replacement: string
) => {
  const before = original.substring(0, startInclusive)
  const after = original.substring(endExclusive)
  return before + replacement + after
}

/**
 * Attempts to assert that the child element is of a supported type. This can't be enforced
 * by the type system so it has to be done as a runtime check. This isn't foolproof - a
 * component that forwards a ref to a correct element but does not forward event handlers
 * will not work. But it's the best we can reasonably do.
 */
export function requireChildrenToBeInput(
  child: React.ReactElement,
  childRef: React.RefObject<HTMLElement>
): TextInputCompatibleChild {
  Children.only(child) // Assert that the child is lonely
  if (
    // There is no way to know what type the underlying child is until it mounts, so this
    // will always pass on first render before failing on the second render
    childRef.current &&
    !(childRef.current instanceof HTMLInputElement) &&
    !(childRef.current instanceof HTMLTextAreaElement)
  ) {
    throw new TypeError(
      `AutocompleteTextarea child must be a component that forwards a ref and props to an <input> or <textarea> element.`
    )
  }
  return child
}

/**
 * Combine several event handlers into one. The last handler in the list is called first
 * and no further handlers will be called if `event.preventDefault()` is called.
 */
export const augmentHandler =
  <E extends SyntheticEvent>(...handlers: Array<EventHandler<E> | undefined>) =>
  (event: E) => {
    for (const handler of [...handlers].reverse()) {
      if (!event.isDefaultPrevented()) handler?.(event)
    }
  }
