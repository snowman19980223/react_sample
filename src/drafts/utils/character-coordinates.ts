import getCaretCoordinates from '@koddsson/textarea-caret'

export type Coordinates = {
  top: number
  left: number
}

/**
 * Obtain the coordinates (px) of the bottom left of a character in an input, relative to the
 * top-left corner of the input itself.
 * @param input The target input element.
 * @param index The index of the character to calculate for.
 * @param adjustForScroll Control whether the returned value is adjusted based on scroll position.
 */
export const getCharacterCoordinates = (
  input: HTMLTextAreaElement | HTMLInputElement | null,
  index: number,
  adjustForScroll = true
): Coordinates => {
  if (!input) return {top: 0, left: 0}

  // word-wrap:break-word breaks the getCaretCoordinates calculations (a bug), and word-wrap has
  // no effect on input element anyway
  if (input instanceof HTMLInputElement) input.style.wordWrap = ''

  let coords = getCaretCoordinates(input, index)

  // The library calls parseInt on the computed line-height of the element, failing to account for
  // the possibility of it being 'normal' (another bug). In that case, fall back to a rough guess
  // of 1.2 based on MDN: "Desktop browsers use a default value of roughly 1.2".
  if (isNaN(coords.height)) coords.height = parseInt(getComputedStyle(input).fontSize) * 1.2

  // Sometimes top is negative, incorrectly, because of the wierd line-height calculations around
  // border-box sized single-line inputs.
  coords.top = Math.abs(coords.top)

  // For some single-line inputs, the rightmost character can be accidentally wrapped even with the
  // wordWrap fix above. If this happens, go back to the last usable index
  let adjustedIndex = index
  while (input instanceof HTMLInputElement && coords.top > coords.height) {
    coords = getCaretCoordinates(input, --adjustedIndex)
  }

  const scrollTopOffset = adjustForScroll ? -input.scrollTop : 0
  const scrollLeftOffset = adjustForScroll ? -input.scrollLeft : 0

  return {top: coords.top + coords.height + scrollTopOffset, left: coords.left + scrollLeftOffset}
}

/**
 * Obtain the coordinates of the bottom left of a character in an input relative to the top-left
 * of the page.
 * @param input The target input element.
 * @param index The index of the character to calculate for.
 */
export const getAbsoluteCharacterCoordinates = (
  input: HTMLTextAreaElement | HTMLInputElement | null,
  index: number
): Coordinates => {
  const {top: relativeTop, left: relativeLeft} = getCharacterCoordinates(input, index, true)
  const {top: viewportOffsetTop, left: viewportOffsetLeft} = input?.getBoundingClientRect() ?? {top: 0, left: 0}

  return {
    top: viewportOffsetTop + relativeTop,
    left: viewportOffsetLeft + relativeLeft
  }
}
