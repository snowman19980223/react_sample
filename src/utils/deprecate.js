/* eslint-disable no-console */

export default function({name, message, version}){
  if (process.env.NODE_ENV !== "production") {
    console.warn(`WARNING! ${name} is deprecated and will be removed in version ${version}. ${message}`)
  }
}
