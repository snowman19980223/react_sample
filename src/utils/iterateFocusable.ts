/**
 * Returns an iterator over all of the focusable elements within `container`.
 * Note: If `container` is itself focusable it will be included in the results.
 * @param container The container over which to find focusable elements.
 * @param reverse If true, iterate backwards through focusable elements.
 */
export function* iterateFocusableElements(
  container: Element,
  reverse = false
): Generator<HTMLElement, undefined, never> {
  const walker = document.createTreeWalker(container, NodeFilter.SHOW_ELEMENT, {
    acceptNode: node =>
      node instanceof HTMLElement && node.tabIndex === -1 ? NodeFilter.FILTER_SKIP : NodeFilter.FILTER_ACCEPT
  })
  let nextNode: Node | null = null;

  // If iterating in reverse, continue traversing down into the last child until we reach
  // a leaf DOM node
  if (reverse) {
      let lastChild = walker.lastChild()
      while (lastChild) {
        nextNode = lastChild
        lastChild = walker.lastChild()
      }
  } else {
    nextNode = walker.firstChild()
  }
  while (nextNode instanceof HTMLElement) {
    yield nextNode
    nextNode = reverse ? walker.previousNode() : walker.nextNode()
  }
  return undefined
}
