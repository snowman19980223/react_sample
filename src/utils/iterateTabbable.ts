/**
 * Options to the tabbable elements iterator
 */
export interface IterateTabbableOptions {
  /**
   * (Default: false) Iterate through tabbable elements in reverse-order
   */
  reverse?: boolean

  /**
   * (Default: false) Perform additional checks to determine tabbability
   * which may adversely affect app performance.
   */
  strict?: boolean
}

/**
 * Returns an iterator over all of the tabbable elements within `container`.
 * Note: If `container` is itself tabbable it will be included in the results.
 * @param container The container over which to find tabbable elements.
 * @param reverse If true, iterate backwards through tabbable elements.
 */
export function* iterateTabbableElements(
  container: Element,
  options: IterateTabbableOptions = {}
): Generator<HTMLElement, undefined, never> {
  const walker = document.createTreeWalker(container, NodeFilter.SHOW_ELEMENT, {
    acceptNode: node =>
      node instanceof HTMLElement && isTabbable(node, options.strict ?? false)
        ? NodeFilter.FILTER_ACCEPT
        : NodeFilter.FILTER_SKIP
  })
  let nextNode: Node | null = null

  // If iterating in reverse, continue traversing down into the last child until we reach
  // a leaf DOM node
  if (options.reverse) {
    let lastChild = walker.lastChild()
    while (lastChild) {
      nextNode = lastChild
      lastChild = walker.lastChild()
    }
  } else {
    nextNode = walker.firstChild()
  }
  while (nextNode instanceof HTMLElement) {
    yield nextNode
    nextNode = options.reverse ? walker.previousNode() : walker.nextNode()
  }
  return undefined
}

/**
 * Determines whether the given element is tabbable. If `strict` is true, we may
 * perform additional checks that require a reflow (less performant)
 * @param elem
 * @param strict
 */
export function isTabbable(elem: HTMLElement, strict = false): boolean {
  const tabIndexInert = elem.tabIndex < 0
  const disabledAttrInert = elem.getAttribute('disabled') != null
  const hiddenInert = elem.hidden
  const hiddenInputInert = elem instanceof HTMLInputElement && elem.type === 'hidden'

  if (tabIndexInert || disabledAttrInert || hiddenInert || hiddenInputInert) {
    return false
  }

  if (strict) {
    const sizeInert = elem.offsetWidth === 0 || elem.offsetHeight === 0
    const visibilityInert = getComputedStyle(elem).visibility === 'hidden'
    const clientRectsInert = elem.getClientRects().length === 0;
    if (sizeInert || visibilityInert || clientRectsInert) {
      return false
    }
  }

  return true
}
