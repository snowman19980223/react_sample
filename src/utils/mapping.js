import PropTypes from 'prop-types'
import classnames from 'classnames'
import {breakpoints} from '../props'
import theme from '../theme'

const {colors, fontSizes, radii} = theme
const {bg: bgColors, border: borderColors, ...namedColors} = colors

const identity = d => d

export function mapPropToClassnames(prop, mapValue) {
  return props => {
    if (prop in props) {
      const {[prop]: value, className, ...rest} = props
      return {
        className: classnames(className, mapValue(value, {key: value})),
        ...rest
      }
    }
    return props
  }
}

export function responsive(mapValue) {
  return (value, data) => {
    return Array.isArray(value)
      ? value
          .map((v, i) => ({value: v, brk: breakpoints[i]}))
          .filter(d => d.value !== null)
          .map(({value, brk}) => {
            return mapValue(value, {...data, brk: brk && `-${brk}`, key: value})
          })
      : mapValue(value, {...data, key: value})
  }
}

export const bg = mapPropToClassnames('bg', key => `bg-${key}`)
bg.propTypes = {bg: PropTypes.oneOf(Object.keys(bgColors))}

export const borderColor = mapPropToClassnames('borderColor', key => `border-${key}`)
borderColor.propTypes = {borderColor: PropTypes.oneOf(Object.keys(borderColors))}

export const color = mapPropToClassnames('color', key => {
  const k = Array.isArray(namedColors[key]) ? `${key}.5` : key
  return `color-${k.replace(/\./g, '-')}`
})
const colorNames = Object.keys(namedColors).concat(getNestedKeys(namedColors))
color.propTypes = {color: PropTypes.oneOf(colorNames)}

export const fontSize = mapPropToClassnames('fontSize', responsive(expander('f{brk}-{key}')))
fontSize.propTypes = {fontSize: PropTypes.oneOf(range(0, fontSizes.length - 1))}

export const borderRadius = mapPropToClassnames('borderRadius', responsive(expander('r{brk}-{key}')))
borderRadius.propTypes = {borderRadius: PropTypes.oneOf(range(0, radii.length - 1))}

function getNestedKeys(obj) {
  return Object.keys(obj).reduce((list, key) => {
    const value = obj[key]
    if (Array.isArray(value)) {
      list.push(...value.map((sub, i) => [key, i].join('.')))
    }
    /* else if (value && typeof value === 'object') {
      list.push(...Object.keys(value).map(sub => [key, sub].join('.')))
    } */
    return list
  }, [])
}

function expander(template) {
  return (value, subs) => {
    return template.replace(/{(\w+)}/g, (_, key) => {
      return defined(subs[key], '')
    })
  }
}

function defined(d, fallback) {
  return d === null || typeof d === 'undefined' ? fallback : d
}

function range(start, end) {
  const numbers = []
  for (let i = start; i <= end; i++) {
    numbers.push(i)
  }
  return numbers
}
