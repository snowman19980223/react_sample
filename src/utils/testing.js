import renderer from 'react-test-renderer'
import enzyme from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'

enzyme.configure({adapter: new Adapter()})

/**
 * Render the component (a React.createElement() or JSX expression)
 * into its intermediate object representation with 'type',
 * 'props', and 'children' keys
 *
 * The returned object can be matched with expect().toEqual(), e.g.
 *
 * ```js
 * expect(render(<Foo />)).toEqual(render(<div foo='bar' />))
 * ```
 */
export function render(component) {
  return renderer.create(component).toJSON()
}

/**
 * Get the HTML class names rendered by the component instance
 * as an array.
 *
 * ```js
 * expect(renderClasses(<div className='a b' />))
 *   .toEqual(['a', 'b'])
 * ```
 */
export function renderClasses(component) {
  const {
    props: {className}
  } = render(component)
  return className ? className.trim().split(' ') : []
}

/**
 * console.error() is ugly af in jest; mock it with a noop,
 * and return the mock so that you can restore it:
 *
 * ```js
 * const mock = silenceConsoleError(jest)
 * // do stuff that calls console.error()
 * mock.mockRestore() // restore console.error()
 * ```
 *
 * Note: we need to pass `jest` as a reference here because
 * jest's mocking functions are bound to a module at runtime, and
 * need to be called as methods rather than pure functions
 * exported from the jest-mock module.
 */
export function silenceConsoleError(jest) {
  return jest.spyOn(console, 'error').mockImplementation(jest.fn())
}

export function mount(component) {
  return enzyme.mount(component)
}
