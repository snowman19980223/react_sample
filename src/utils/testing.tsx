import React from 'react'
import {promisify} from 'util'
import renderer from 'react-test-renderer'
import enzyme from 'enzyme'
import Adapter from '@wojtekmaj/enzyme-adapter-react-17'
import {ThemeProvider} from 'styled-components'
import {default as defaultTheme} from '../theme'

// eslint-disable-next-line @typescript-eslint/no-var-requires
const readFile = promisify(require('fs').readFile)

export const COMPONENT_DISPLAY_NAME_REGEX = /^[A-Z][A-Za-z]+(\.[A-Z][A-Za-z]+)*$/

enzyme.configure({adapter: new Adapter()})

export function mount(component: React.ReactElement) {
  return enzyme.mount(component)
}

declare global {
  // eslint-disable-next-line @typescript-eslint/no-namespace
  namespace jest {
    interface Matchers<R> {
      toImplementSystemProps: (systemProps: any) => boolean
      toImplementSxProp: () => boolean
      toImplementSxBehavior: () => boolean
      toSetDefaultTheme: () => boolean
      toSetExports: (exports: Record<string, string>) => boolean
    }
  }
}

/**
 * Render the component (a React.createElement() or JSX expression)
 * into its intermediate object representation with 'type',
 * 'props', and 'children' keys
 *
 * The returned object can be matched with expect().toEqual(), e.g.
 *
 * ```js
 * expect(render(<Foo />)).toEqual(render(<div foo='bar' />))
 * ```
 */
export function render(component: React.ReactElement) {
  return renderer.create(component).toJSON() as renderer.ReactTestRendererJSON
}

/**
 * Render the component (a React.createElement() or JSX expression)
 * using react-test-renderer and return the root node
 * ```
 */
export function renderRoot(component: React.ReactElement) {
  return renderer.create(component).root
}

/**
 * Get the HTML class names rendered by the component instance
 * as an array.
 *
 * ```js
 * expect(renderClasses(<div className='a b' />))
 *   .toEqual(['a', 'b'])
 * ```
 */
export function renderClasses(component: React.ReactElement): string {
  const {
    props: {className}
  } = render(component)
  return className ? className.trim().split(' ') : []
}

/**
 * Returns true if a node renders with a single class.
 */
export function rendersClass(node: React.ReactElement, klass: string): boolean {
  return renderClasses(node).includes(klass)
}

export function renderWithTheme(node: React.ReactElement, theme = defaultTheme): renderer.ReactTestRendererJSON {
  return render(<ThemeProvider theme={theme}>{node}</ThemeProvider>)
}

export function px(value: number | string): string {
  return typeof value === 'number' ? `${value}px` : value
}

export function percent(value: number | string): string {
  return typeof value === 'number' ? `${value}%` : value
}

export function renderStyles(node: React.ReactElement): any {
  const {
    props: {className}
  } = render(node)
  return getComputedStyles(className)
}

export function getComputedStyles(className: string): Record<string, string | Record<string, string>> {
  const div = document.createElement('div')
  div.className = className

  const computed = {} as any
  for (const sheet of document.styleSheets) {
    // CSSRulesLists assumes every rule is a CSSRule, not a CSSStyleRule
    for (const rule of sheet.cssRules) {
      if (rule instanceof CSSMediaRule) {
        readMedia(rule)
      } else if (rule instanceof CSSStyleRule) {
        readRule(rule, computed)
      } else {
        // console.warn('rule.type =', rule.type)
      }
    }
  }

  return computed

  function matchesSafe(node: HTMLDivElement, selector: string) {
    if (!selector) {
      return false
    }
    try {
      return node.matches(selector)
    } catch (error) {
      return false
    }
  }

  function readRule(rule: CSSStyleRule, dest: any) {
    if (matchesSafe(div, rule.selectorText)) {
      const {style} = rule
      for (let i = 0; i < style.length; i++) {
        const prop = style[i]
        dest[prop] = style.getPropertyValue(prop)
      }
    } else {
      // console.warn('no match:', rule.selectorText)
    }
  }

  function readMedia(mediaRule: CSSMediaRule) {
    const key = `@media ${mediaRule.media[0]}`
    // const dest = computed[key] || (computed[key] = {})
    const dest = {}
    for (const rule of mediaRule.cssRules) {
      if (rule instanceof CSSStyleRule) {
        readRule(rule, dest)
      }
    }

    // Don't add media rule to computed styles
    // if no styles were actually applied
    if (Object.keys(dest).length > 0) {
      computed[key] = dest
    }
  }
}
/**
 * This provides a layer of compatibility between the render() function from
 * react-test-renderer and Enzyme's mount()
 */
export function getProps(node: React.ReactElement) {
  return typeof node.props === 'function' ? node.props() : node.props
}

export function getClassName(node: React.ReactElement) {
  return getProps(node).className
}

export function getClasses(node: React.ReactElement) {
  const className = getClassName(node)
  return className ? className.trim().split(/ +/) : []
}

export async function loadCSS(path: string) {
  const css = await readFile(require.resolve(path), 'utf8')
  const style = document.createElement('style')
  style.setAttribute('data-path', path)
  style.textContent = css
  document.head.appendChild(style)
  return style
}

export function unloadCSS(path: string) {
  const style = document.querySelector(`style[data-path="${path}"]`)
  if (style) {
    style.remove()
    return true
  }
}

// If a component requires certain props or other conditions in order
// to render without errors, you can pass a `toRender` function that
// returns an element ready to be rendered.

interface Options {
  skipAs?: boolean
  skipSx?: boolean
}

interface BehavesAsComponent {
  Component: React.ComponentType<any>
  systemPropArray: any[]
  toRender?: () => React.ReactElement
  options?: Options
}

export function behavesAsComponent({Component, systemPropArray, toRender, options}: BehavesAsComponent) {
  options = options || {}

  const getElement = () => (toRender ? toRender() : <Component />)

  if (!options.skipSx) {
    it('implements sx prop behavior', () => {
      expect(getElement()).toImplementSxBehavior()
    })
  }

  if (!options.skipAs) {
    it('respects the as prop', () => {
      const As = React.forwardRef<HTMLDivElement>((_props, ref) => <div className="as-component" ref={ref} />)
      const elem = React.cloneElement(getElement(), {as: As})
      expect(render(elem)).toEqual(render(<As />))
    })
  }

  it('sets a valid displayName', () => {
    expect(Component.displayName).toMatch(COMPONENT_DISPLAY_NAME_REGEX)
  })

  it('sets the default theme', () => {
    expect(getElement()).toSetDefaultTheme()
  })

  it('renders consistently', () => {
    expect(render(getElement())).toMatchSnapshot()
  })
}

export function checkExports(path: string, exports: Record<any, any>): void {
  it('has declared exports', () => {
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    const mod = require(`../${path}`)
    expect(mod).toSetExports(exports)
  })
}
