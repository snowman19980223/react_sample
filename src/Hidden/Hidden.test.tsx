import React from 'react'
import '@testing-library/jest-dom/extend-expect'
import {render} from '@testing-library/react'
import {Hidden} from '.'
import MatchMediaMock from 'jest-matchmedia-mock'
import {behavesAsComponent, checkExports, renderStyles, checkStoriesForAxeViolations} from '../utils/testing'
import {mediaQueries} from '../utils/layout'

let matchMedia: MatchMediaMock
describe('Hidden', () => {
  beforeAll(() => {
    matchMedia = new MatchMediaMock()
  })

  afterAll(() => {
    matchMedia.clear()
  })

  behavesAsComponent({
    Component: Hidden,
    options: {skipAs: true, skipSx: true},
    toRender: () => (
      <Hidden on={'narrow'}>
        <div>Hidden on narrow</div>
      </Hidden>
    ),
  })

  checkExports('Hidden', {
    default: Hidden,
    Hidden,
  })

  it('renders `on` prop as expected', () => {
    const {container} = render(
      <Hidden on={'narrow'}>
        <div>Hidden on narrow</div>
      </Hidden>,
    )
    expect(container).toMatchSnapshot()
  })

  it('renders the styles as expected when a single viewport value is provided as a string via `on` prop', () => {
    const expectedStyles = {
      // `.replace` is used because renderStyles return the JSON object without a space after the column
      [`${mediaQueries.regular.replace(': ', ':')}`]: {
        display: 'none',
      },
    }
    expect(
      renderStyles(
        <Hidden on="regular">
          <div>This is hidden on regular viewports</div>
        </Hidden>,
      ),
    ).toEqual(expect.objectContaining(expectedStyles))
  })

  it('renders the styles as expected when multiple viewport values are provided as an array via `on` prop', () => {
    const expectedStyles = {
      [`${mediaQueries.narrow.replace(': ', ':')}`]: {
        display: 'none',
      },
      [`${mediaQueries.wide.replace(': ', ':')}`]: {
        display: 'none',
      },
    }
    expect(
      renderStyles(
        <Hidden on={['narrow', 'wide']}>
          <div>This is hidden on regular and wide viewports</div>
        </Hidden>,
      ),
    ).toEqual(expect.objectContaining(expectedStyles))
  })
})

checkStoriesForAxeViolations('features', '../Hidden/')
checkStoriesForAxeViolations('examples', '../Hidden/')
