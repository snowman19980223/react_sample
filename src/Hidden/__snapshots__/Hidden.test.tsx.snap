// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Hidden renders \`on\` prop as expected 1`] = `
@media screen and (max-width:calc(768px - 0.02px)) {
  .c0 {
    display: none;
  }
}

<div>
  <div
    class="c0"
  >
    <div>
      Hidden on narrow
    </div>
  </div>
</div>
`;

exports[`Hidden renders consistently 1`] = `
@media screen and (max-width:calc(768px - 0.02px)) {
  .c0 {
    display: none;
  }
}

<div
  className="c0"
>
  <div>
    Hidden on narrow
  </div>
</div>
`;
